File: 863.All Nodes Distance K in Binary Tree/answer.cpp
class Solution {
 public:
  vector<int> distanceK(TreeNode* root, TreeNode* target, int K) {
    set<int> vals;
    map<TreeNode*, vector<TreeNode*>> adj;
    function<void(TreeNode*)> dfs = [&](TreeNode* at) {
      if (at->left) {
        adj[at].push_back(at->left);
        adj[at->left].push_back(at);
        dfs(at->left);
      }
      if (at->right) {
        adj[at].push_back(at->right);
        adj[at->right].push_back(at);
        dfs(at->right);
      }
    };
    dfs(root);
    queue<TreeNode*> que;
    map<TreeNode*, ll> dist;
    dist[target] = 0;
    que.push(target);
    while (not que.empty()) {
      auto at = que.front();
      que.pop();
      if (dist[at] == K) {
        vals.insert(at->val);
      }
      RARA(to, adj[at]) if (not dist.count(to)) {
          dist[to] = dist[at] + 1;
          que.push(to);
        }
    }
    return vector<ll>(ALL(vals));
  }
};

File: 863.All Nodes Distance K in Binary Tree/answer.py
def distanceK(self, root, target, K):
    conn = collections.defaultdict(list)
    def connect(parent, child):
        if parent and child:
            conn[parent.val].append(child.val)
            conn[child.val].append(parent.val)
        if child.left: connect(child, child.left)
        if child.right: connect(child, child.right)
    connect(None, root)
    bfs = [target.val]
    seen = set(bfs)
    for i in xrange(K):
        bfs = [y for x in bfs for y in conn[x] if y not in seen]
        seen |= set(bfs)
    return bfs
